Issue 14706 webui

spylogsster opened the pull request:
<!-- Add brave-browser issue bellow that this PR will resolve -->
Resolves 

## Submitter Checklist:

- [ ] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

petemill commented on components/trezor_bridge/resources/app.tsx:
```diff
@@ -0,0 +1,82 @@
+})
+
+function initialize () {
+  render(<div/>, document.getElementById('root'))
```
This isn't needed. Or are you planning on adding a UI later on?

spylogsster replied:
done

petemill commented on components/trezor_bridge/resources/trezor/data/bridge/releases.json:
```diff
@@ -0,0 +1,41 @@
+[
```
Since the entire data/ directory is copied from the published `trezor-connect` npm module, it is already available to the source code via `node_modules/trezor-connect/data`. Therefore we do not need to commit this large amount of data to git.
Instead, you can use GN `copy` template to copy each file before packaging using the same grit .grd file, e.g.

```gn
data_dir = "//brave/node_modules/trezor-connect/data"

copy("data") {
  visibility = [ ":*" ]
  sources = [
    "$data_dir/bridge/latest.txt",
    "$data_dir/bridge/releases.json",
    "$data_dir/bridge/coins.json",
    "$data_dir/bridge/config.json",
    ...
    "$data_dir/bridge/messages/messagesv7.json",
    ...
  ]
  outputs = [ "$root_gen_dir/components/trezor_bridge/resources/trezor/data/{{source_file_part}}" ]
}
```

spylogsster replied:
done

petemill commented on components/trezor_bridge/resources/app.tsx:
```diff
@@ -0,0 +1,82 @@
+
+// Hooks to trick some checks inside TrezorConnect to use webextension env.
+function setupBraveHooks () {
+  if (!window.chrome) {
+    window.chrome = Object({})
+  }
+  if (!window.chrome.runtime) {
+    window.chrome.runtime = Object({})
+  }
+  if (!window.chrome.runtime.id) {
+    window.chrome.runtime.id = ''
+  }
+  if (!window.chrome.runtime.onConnect) {
+    window.chrome.runtime.onConnect = Object({
+      addListener: () => undefined
+    })
+  }
```
hmm, I'm not sure why we would need to stub these APIs if we're using the browser version of trezor-connect and not the webextension version...

spylogsster replied:
we are using the webextension version, it doesnt have `browser` version: `"node" | "web" | "webextension" | "electron" | "react-native" | undefined`

spylogsster replied:
changed to `web`

petemill commented on components/trezor_bridge/resources/app.tsx:
```diff
@@ -0,0 +1,82 @@
+    email: 'support@brave.com',
+    appUrl: 'web-ui-boilerplate'
+  },
+  env: 'webextension'
```
should this be browser?

spylogsster replied:
`Type '"browser"' is not assignable to type '"node" | "web" | "webextension" | "electron" | "react-native" | undefined'.`

petemill replied:
`'web'`, then?

spylogsster replied:
yep, done

petemill commented on components/trezor_bridge/resources/app.tsx:
```diff
@@ -0,0 +1,82 @@
+// Copyright (c) 2021 The Brave Authors. All rights reserved.
```
Lets match this file name with the output filename as well as the html file name, otherwise it can be confusing - in your BUILD.gn and in the htmll you reference this as `trezor_bridge`, but here it is called `app` and the html file is called `bridge`.

spylogsster replied:
done

petemill commented on components/trezor_bridge/resources/trezor/iframe.tsx:
```diff
@@ -0,0 +1,35 @@
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// you can obtain one at http://mozilla.org/MPL/2.0/.
+
+import * as React from 'react'
```
If this can be removed (see comment about UI, below), then we should rename this file from `iframe.tsx` to `iframe.ts`

spylogsster replied:
done

petemill commented on components/trezor_bridge/resources/trezor/iframe.tsx:
```diff
@@ -0,0 +1,35 @@
+
+function initialize () {
+  render(<div/>, document.getElementById('root'))
+
```
This won't be needed, since `trezor-connect` creates its own UI via DOM API, right?

spylogsster replied:
yep, removed

petemill commented on components/resources/brave_components_strings.grd:
```diff
@@ -1032,6 +1032,7 @@
       <part file="decentralized_dns_strings.grdp" />
       <part file="domain_block_strings.grdp" />
       <part file="ipfs_strings.grdp" />
+      <part file="trezor_bridge_strings.grdp" />
```
nit: I think we should not keep adding to this file, but make a unique one for the trezor-bridge component. Plus, this needs a conditional for if wallet is enabled.

petemill commented on components/resources/trezor_bridge_resources.grdp:
```diff
@@ -0,0 +1,14 @@
+  <include name="IDR_TREZOR_BRIDGE_IFRAME" file="../trezor_bridge/resources/trezor/iframe.html" type="BINDATA" />
+  <include name="IDR_TREZOR_BRIDGE_IFRAME_DATA_CONFIG" file="../trezor_bridge/resources/trezor/data/config.json" type="BINDATA" />
+  <include name="IDR_TREZOR_BRIDGE_IFRAME_DATA_COINS" file="../trezor_bridge/resources/trezor/data/coins.json" type="BINDATA" />
+  <include name="IDR_TREZOR_BRIDGE_IFRAME_DATA_MESSAGES_MESSAGES" file="../trezor_bridge/resources/trezor/data/messages/messages.json" type="BINDATA" />
```
If you follow my advice on using GN `copy` to move the files here, then you would change these references to the generated directory

spylogsster replied:
done

petemill commented on components/trezor_bridge/trezor_bridge_handler.cc:
```diff
@@ -0,0 +1,184 @@
+void TrezorBridgeHandler::HandleFetchRequest(const base::ListValue* args) {
+  std::string url;
+  CHECK(args->GetString(1, &url));
+  if (trezor_data_resources.contains(url)) {
+    auto resource_id = trezor_data_resources.at(url);
+    const ui::ResourceBundle& resource_bundle =
+        ui::ResourceBundle::GetSharedInstance();
+    const std::string& resource_text =
+        resource_bundle.LoadDataResourceString(resource_id);
+    RespondRequestCallback(args->GetList()[0], true, resource_text, "ok");
+    return;
+  }
```
I think you can simplify and remove this part by adding those resources to the data source in the same way as the html files, and we can lean on regular webui data source to serve the large data. And then also filter out relative URLs from your `window.fetch` JS override.

If you create a GRD just for your component then iterate over the c++ generated map to add the source, you can control the relative url they will be served from, and you can provide a hint for yourself with the `resource_path` attribute, if neccessary:

```xml
<include name="IDR_TREZOR_BRIDGE_IFRAME_DATA_MESSAGES_MESSAGES_V6" file="${target_gen_dir}/messages/messages_v6.json" resource_path="messages/messages_v6.json" use_base_dir="false" type="BINDATA"/>
```

spylogsster replied:
doesnt work, the library uses window.fetch for json file

petemill replied:
Yes and I'm suggesting to clean up our workarounds by allowing it to fetch as normal for those resources, so instead of re-inventing the resource serving here, we simply add them to the data source, and if you make the resources in their own GRD file and grit target as suggested in my other comment, then you would have something like:

```cpp
  for (size_t i = 0; i < kTrezorBridgeStaticResourcesSize; ++i) {
    source->AddResourcePath(kTrezorBridgeStaticResources[i].resource_path, kTrezorBridgeStaticResources[i].id);
  }
```

petemill replied:
We just discovered this won't work because `window.fetch` does not support the `chrome` url scheme. Other workaround would be to use webpack to import the assets and avoid all the gn file copy and grit GRD etc, but the way you have it currently, serving the assets manually from the c++ message handler, is fine too IMO 👍

spylogsster replied:
doesnt work, discussed in slack

petemill commented on package.json:
```diff
@@ -301,6 +301,9 @@
     "style-loader": "0.23.1",
     "styled-components": "5.3.0",
+    "trezor-connect": "8.1.5-extended",
+    "sharedworker-loader": "^2.1.1",
+    "worker-loader": "2.0.0",
```
Please add a comment as to why we needed to add these even though they aren't directly referenced by our code. Also remove the `^` prefix, as we seem to specify exact versions for all dependencies.

spylogsster replied:
done

petemill replied:
I don't see the comment, how about adding the reason to the git commit message for this PR? That would be helpful for me. Otherwise I'd see the dependency, see that it's not directly used and remove it.

spylogsster replied:
done

petemill reviewed (changes requested):
Nice to have worked with you on this! Please also request a security review for the webui CSP overrides (although I think they're fine), for the new packages, and the general architecture of rendering trezor-connect-provided dynamic UI in webui.

petemill commented on components/resources/brave_components_resources.grd:
```diff
@@ -69,6 +69,7 @@
       <part file="speedreader_resources.grdp" />
       <part file="brave_flags_ui_resources.grdp" />
       <part file="ipfs_resources.grdp" />
+      <part file="trezor_bridge_resources.grdp" />
```
This is the one I think we should add in the trezor-bridge component, not this shared component, please

petemill replied:
we can do this as a follow-up since I see we're still adding individual component resources to this shared .grd file, e.g. wallet and vpn. However, I think you might need to wrap with a condition, e.g. 
```xml
<if expr="not is_android"
```
?

spylogsster replied:
done

petemill commented on components/trezor_bridge/BUILD.gn:
```diff
@@ -0,0 +1,12 @@
+    "//brave/components/resources:strings_grit",
+    "//content/public/browser",
+  ]
+}
```
You would add the grit target for trezor_bridge_resources to this BUILD.gn

spylogsster replied:
done

petemill commented on components/trezor_bridge/resources/BUILD.gn:
```diff
@@ -0,0 +1,39 @@
+  resource_name = "trezor_bridge"
+  output_dir = "$root_gen_dir/brave/components/trezor_bridge/resources"
+  deps = [
+    ":data",
```
I don't think this is right - this target doesn't depend on `:data` because it's not the one to include the data files. The grit target which uses the copied files is the one which should depend on it.

spylogsster replied:
done

petemill commented on components/trezor_bridge/resources/BUILD.gn:
```diff
@@ -0,0 +1,39 @@
+
+node_data_dir = "//brave/node_modules/trezor-connect/data"
+
+copy("data") {
```
This target should probably be moved to `trezor_bridge/BUILD.gn` as that's where the message handler that uses the resources is.

spylogsster replied:
done

petemill commented on components/trezor_bridge/resources/trezor/iframe.ts:
```diff
@@ -0,0 +1,26 @@
+
+// Overriding window.fetch to catch network requests in the page handler.
+window.fetch = function (url, options) {
+  return new Promise(async (resolve, reject) => {
```
This is where you can exclude regular relative requests, e.g. `/data/messages` or `/another.webpack.chunk.js` so that we can reduce this patch to only proxy requests to the trezor service.

```ts
if (url.startsWith('/') {
  return nativeWindowFetch(url, options)
}
```

spylogsster replied:
doesnt work. discussed in slack

petemill commented on components/trezor_bridge/resources/trezor_bridge.html:
```diff
@@ -0,0 +1,10 @@
+  <title>Trezor Bridge</title>
+</head>
+<body>
+  <div id="root"></div>
```
This element is not needed

spylogsster replied:
done

petemill commented on components/trezor_bridge/resources/trezor_bridge.ts:
```diff
@@ -0,0 +1,74 @@
+    email: 'support@brave.com',
+    appUrl: 'web-ui-boilerplate'
+  },
+  env: 'webextension'
```
If you use `'web'` does it remove the need to polyfill `chrome.runtime` and everything still works?

spylogsster replied:
done

petemill commented:
Please squash the commits so we don't push all those JSON file bytes to git.

petemill reviewed (changes requested):
Thanks for addressing all the feedback! One more issue I see is that the conditions for including targets are different in 3 different ways:

browser/ui includes the trezor bridge webui onlly if:
```
if (brave_wallet_enabled && !is_android && !is_ios) {
```

components/resources/build.gn requests trezor-bridge ui to be built only if
```
if (brave_wallet_enabled && !is_ios) {
```

but componentss/resources/brave_components__{resources,strings}.grd includes the resources:
```
<if expr="not is_android">
```

If these are to be desktop-only and wallet-only then perhaps change them all to be `(brave_wallet_enabled && !is_android && !is_ios)`

spylogsster commented:
> Thanks for addressing all the feedback! One more issue I see is that the conditions for including targets are different in 3 different ways:
> 
> browser/ui includes the trezor bridge webui onlly if:
> 
> ```
> if (brave_wallet_enabled && !is_android && !is_ios) {
> ```
> 
> components/resources/build.gn requests trezor-bridge ui to be built only if
> 
> ```
> if (brave_wallet_enabled && !is_ios) {
> ```
> 
> but componentss/resources/brave_components__{resources,strings}.grd includes the resources:
> 
> ```
> <if expr="not is_android">
> ```
> 
> If these are to be desktop-only and wallet-only then perhaps change them all to be `(brave_wallet_enabled && !is_android && !is_ios)`

for now it is enough to keep it under brave_wallet_enabled, wrapped strings/resources in last commit, thanks for checking this 👍

diracdeltas commented on browser/ui/webui/brave_wallet/trezor_bridge_ui.cc:
```diff
@@ -0,0 +1,50 @@
+      network::mojom::CSPDirectiveName::ScriptSrc,
+      "script-src chrome://resources/ chrome://trezor-bridge ;");
+  // Disable DenyXFrame to allow TrezorConnect to create iframes.
+  html_source->DisableDenyXFrameOptions();
```
i suggest setting the `frame-ancestors` CSP directive so it can only be framed by whitelisted origins

spylogsster replied:
done, thanks!

thypon replied:
I think @diracdeltas meant (1) add `frame-ancestors` and (2) remove the `html_source->DisableDenyXFrameOptions();`. They should be mutually exclusive.

bridiver commented on browser/resources/resource_ids:
```diff
@@ -121,4 +121,7 @@
   "<(ROOT_GEN_DIR)/brave/web-ui-brave_vpn_panel/brave_vpn_panel.grd": {
     "includes": [51000]
   },
+  "<(ROOT_GEN_DIR)/brave/web-ui-trezor_bridge/trezor_bridge.grd": {
```
We need to stop adding grd files like this. Please create `brave/browser/browser_resources.grd` and make `trezor_bridge` a grdp file

petemill replied:
That's going to be beyond the scope of this PR. The webpack GN template spits out a GRD, so that would need to be changed to GRDP. That's going to affect _a lot_ of things since we depend on a specific `kMyGRDFileResourceMap` being created for each UI, and not 1 map of every resource for the whole browser.

bridiver replied:
ok, I didn't realize this was part of a larger process and couldn't easily be changed

bridiver commented on browser/ui/BUILD.gn:
```diff
@@ -528,7 +528,16 @@ source_set("ui") {
     deps += [ "//brave/components/tor/resources:internals_generated_resources" ]
   }
-
+  if (brave_wallet_enabled && !is_ios) {
```
`!is_ios` should not apply here. ios should never have a dep on `//brave/browser/ui`

bridiver commented on components/trezor_bridge/trezor_bridge_handler.cc:
```diff
@@ -0,0 +1,185 @@
+
+}  // namespace
+
+TrezorBridgeHandler::TrezorBridgeHandler(
```
what is going on here? This is not the correct way to load resources from a pak file. cc @petemill

bridiver commented on components/trezor_bridge/resources/trezor/iframe.ts:
```diff
@@ -0,0 +1,26 @@
+import 'trezor-connect/lib/iframe/iframe'
+const { setFetch } = require('trezor-link/lib/bridge/http')
+
+// Overriding window.fetch to catch network requests in the page handler.
```
https://chromium.googlesource.com/chromium/src/+/HEAD/docs/webui_explainer.md#security-considerations
"WebUI pages cannot issue http/https fetches"
cc @diracdeltas

bridiver commented on components/trezor_bridge/trezor_bridge_handler.cc:
```diff
@@ -0,0 +1,185 @@
+
+void TrezorBridgeHandler::RegisterMessages() {
+  web_ui()->RegisterMessageCallback(
+      "trezor-fetch",
```
this method name does not follow standard (camelcase) naming conventions

bridiver commented on components/trezor_bridge/trezor_bridge_handler.cc:
```diff
@@ -0,0 +1,185 @@
+                          base::Unretained(this)));
+}
+
+void TrezorBridgeHandler::HandleFetchRequest(const base::ListValue* args) {
```
`const base::Value&`

bridiver commented on components/resources/BUILD.gn:
```diff
@@ -96,8 +97,15 @@ repack("resources") {
-    sources += [ "$root_gen_dir/brave/components/brave_wallet/resources/brave_wallet_script_generated.pak" ]
+    deps += [
+      "//brave/components/brave_wallet/resources:ethereum_provider_generated_resources",
+      "//brave/components/trezor_bridge/:copy_trezor_data",
```
`copy_trezor_data` should be a dep internal to `brave/components/trezor_bridge/resources`

bridiver replied:
the output path can be specified independently of the input path
